README.md
# EasyPost PHP Client Library

[![Build Status](https://travis-ci.com/EasyPost/easypost-php.svg?branch=master)](https://travis-ci.com/EasyPost/easypost-php)

EasyPost is a simple shipping API. You can sign up for an account at https://easypost.com.

## Installation

***NOTE:** Although this library has been tested with PHP 5.3^, it's highly recommended to use PHP 7.2^. Additionally, This library relies on the [mbstring](http://php.net/manual/en/book.mbstring.php) extension. Ensure you have it [installed](http://www.php.net/manual/en/mbstring.installation.php) correctly before using the library.*

### Install Client Library

**Install the Library via [Composer](http://getcomposer.org/):**

```shell
composer require easypost/easypost-php
```

Then, require the autoloader:

```php
require_once("/path/to/vendor/easypost/autoload.php");
```

**OR: Manually Install the Library**

Manually download this library and require it from your project:

```php
require_once("/path/to/lib/easypost.php");
```

### Install Dependencies

Run the following from the root directory of the library:

```shell
composer install
```

## Example

```php
require_once("/path/to/vendor/easypost/autoload.php");
\EasyPost\EasyPost::setApiKey('cueqNZUb3ldeWTNX7MU3Mel8UXtaAMUi');

$to_address = \EasyPost\Address::create(
    array(
        "name"    => "Dr. Steve Brule",
        "street1" => "179 N Harbor Dr",
        "city"    => "Redondo Beach",
        "state"   => "CA",
        "zip"     => "90277",
        "phone"   => "310-808-5243"
    )
);
$from_address = \EasyPost\Address::create(
    array(
        "company" => "EasyPost",
        "street1" => "118 2nd Street",
        "street2" => "4th Floor",
        "city"    => "San Francisco",
        "state"   => "CA",
        "zip"     => "94105",
        "phone"   => "415-456-7890"
    )
);
$parcel = \EasyPost\Parcel::create(
    array(
        "predefined_package" => "LargeFlatRateBox",
        "weight" => 76.9
    )
);
$shipment = \EasyPost\Shipment::create(
    array(
        "to_address"   => $to_address,
        "from_address" => $from_address,
        "parcel"       => $parcel
    )
);

$shipment->buy($shipment->lowest_rate());

$shipment->insure(array('amount' => 100));

echo $shipment->postage_label->label_url;
```

## Documentation

Up-to-date documentation can be found at: https://www.easypost.com/docs.

## Releasing

1. Update the `VERSION` file
1. Update the `CHANGELOG` file
1. Update the `VERSION` constant in the `EasyPost.php` file
1. Update the version in the `composer.json` file
1. Create a git tag with the new version

[Packagist](https://packagist.org) should automatically pull in the new tag if the above was done properly.

## Testing & Development

***NOTE: Unit tests only work with PHP 7.2^.***

Ensure dependencies are installed, then run any of the following:

**Fix PHP Standards:**
```shell
./bin/php-cs-fixer fix mydir --verbose --show-progress=estimating     
```

**Run Linting:**
```shell
./bin/phplint ./dir
```

**Run Unit Tests:**
```shell
./bin/phpunit
```


CHANGELOG
=== 3.4.5 2020-05-07

* Added "Refund" and "Shipment Invoice" to the prefix and type lists


=== 3.4.4 2020-03-11

* Version bump with updated author


=== 3.4.3 2020-03-10

* Bumped version in package.json for Packagist
* Updated README with release instructions


=== 3.4.2 2020-03-09

* Fixed a bug that would not create reports properly
* Added unit tests for reports


=== 3.4.1 2018-04-13

* Added ability to set connection and response timeouts for all EasyPost actions (thanks mattjanssen!)


=== 3.4.0 2017-11-30

* Renamed `Object` class `EasyPostObject` since Object is a keyword in PHP7.2


=== 3.3.4 2017-05-24

* Fixed confusing Pickup example
* Enabled Reports to retrieve without a type parameter
* Use X-Client-User-Agent over X-EasyPost-Client-User-Agent


=== 3.3.3 2017-02-14

* Added get_rates method to Order objects


=== 3.3.2 2017-01-24

* Fixed compatibility issues with older PHP versions and the Report class


=== 3.3.1 2017-01-19

* Fixed Create for ScanForms


=== 3.3.0 2017-01-17

* Added basic CRUD methods for Webhooks


=== 3.2.1 2016-12-14

* Fixed error in Util.php they may have affected shipment_reports retrievals


=== 3.2.0 2016-12-12

* Added ability to create Report objects for shipment, tracker and payment_log reports


=== 3.1.3 2016-10-25

* Fixed warning message from Address Verify/VerifyStrict


=== 3.1.2 2016-09-16

* Added create_list for trackers


=== 3.1.1 2016-07-21

* Added Insurance support


=== 3.1.0 2016-06-02

* Renamed "Resource" class to "EasypostResource" since Resource is reserved in PHP7 (thanks ccovey!)


=== 3.0.3 2016-02-22

* Added ability to create top-level users


=== 3.0.2 2016-01-29

* Added 'verify' and 'verify_strict' options to Address creation


=== 3.0.1 2016-01-05

* Updated ReadMe (thanks browner12!)
* Added DocBlock annotations to all methods (thanks browner12!)


=== 3.0.0 2015-12-18

* Fixed Shipment Retrieve by tracking_code


=== 2.1.2 2015-11-04

* Added Tracker.all example, and updated some old examples


=== 2.1.1 2015-07-01

* Removed cURL options for SSL version, timeout, and connect timeout. The
default values for these options are acceptable.


=== 2.1.0 2015-04-10

* Added User and CarrierAccount resources and examples for each.
* Modified EasyPost Object setter to record the change in the top level parent
object rather than the object that the change was made in. This allows you to
change sub objects and then call save() on the parent to have all properties
saved.
* Removed EasyPost Object transientValues and made unsavedValues act as a
representation of the params to send on save, rather than a flag to indicate
if a property had been changed.
* Added support for EasyPost's new error format: {"error": {"message": "...",
"code": ..., "errors": {...}}} as well as a prettyPrint function to better
display errors in development.
* Added support for additional HTTP methods PATCH and PUT.
* EasyPost Objects are now Iterable.


=== 2.0.12 2015-02-04

* Added pickup class
* Added pickup example


=== 2.0.11 2014-11-04

* Added tracker to shipment buy response
* Updated tracker examples


=== 2.0.10 2014-07-07

* Added Item, Container, and Order resources.


=== 2.0.9 2014-04-08

* Updated the CA cert bundle.


=== 2.0.8 2013-12-29

* Added the ability to add negative filters to Shipment->lowest_rate by starting them with an exclamation mark.


=== 2.0.7 2013-11-08

* Added endpoint to create shipment from tracking code.


=== 2.0.6 2013-10-21

* Added Event resource for webhooks.


=== 2.0.5 2013-08-30

* API Addition: Batch.buy method added. Trigger the purchasing of labels after a successful Batch.create.
* API Addition: Batch.scan_form method added. Quickly generate a scan_form for an entire batch.


=== 2.0.4 2013-07-31

* API Addition: Tracker objects added. Trackers can be used to register any tracking code with EasyPost webhooks.


=== 2.0.3 2013-07-16

* Bug fix: error messages should no longer be truncated to their first character.


=== 2.0.2 2013-07-05

* Added function to Address to all creating and verifying at the same time.


=== 2.0.1 2013-07-01

* Add label function to Shipment to request specific label file_formats (pdf, epl2, zpl).
* Add insure function to Shipment. Add insurance to any shipment in one call!


examples/batches.php
<?php

require_once("../lib/easypost.php");

\EasyPost\EasyPost::setApiKey('cueqNZUb3ldeWTNX7MU3Mel8UXtaAMUi');

// create addresses
$from_address = array(
    "company" => "EasyPost",
    "street1" => "388 Townsend St",
    "city"    => "San Francisco",
    "state"   => "CA",
    "zip"     => "94107-8273",
    "phone"   => "415-456-7890"
);
$parcel = array(
    "predefined_package" => 'Parcel',
    "weight"             => 22.9
);
$customs_item = array(
    "description"      => "T Shirt",
    "currency"         => "USD",
    "value"            => 35.0,
    "quantity"         => 7,
    "weight"           => 22.9,
    "hs_tariff_number" => "123456"
);
$customs_info = array(
    "contents_type" => "merchandise",
    "customs_items" => array(
      $customs_item
    )
);

// in your application orders will likely
// come from an external data source
$orders = array(
    array(
        "address"  => array(
            "name"    => "Ronald",
            "street1" => "6258 Amesbury St",
            "city"    => "San Diego",
            "state"   => "CA",
            "zip"     => "92114"
        ),
        "reference"   => "123456786",
        "carrier"     => "USPS",
        "service"     => "Priority"
    ),
    array(
        "address"  => array(
            "name"    => "Hamburgler",
            "street1" => "8308 Fenway Rd",
            "city"    => "Bethesda",
            "state"   => "MD",
            "zip"     => "20817"
        ),
        "reference"   => "123456787",
        "carrier"     => "USPS",
        "service"     => "Priority"
    ),
    array(
        "address"  => array(
            "name"    => "Grimace",
            "street1" => "10 Wall St",
            "city"    => "Burlington",
            "state"   => "MA",
            "zip"     => "01803"
        ),
        "reference"   => "123456788",
        "carrier"     => "USPS",
        "service"     => "Priority"
    ),
    array(
        "address"  => array(
            "name"    => "Cosmc",
            "street1" => "3315 W Greenway Rd",
            "city"    => "Phoenix",
            "state"   => "AZ",
            "zip"     => "85053"
        ),
        "reference"   => "123456789",
        "carrier"     => "USPS",
        "service"     => "Express"
    ),
    array(
        "address"      => array(
            "company" => "Tim Hortons",
            "street1" => "65 Queen St W",
            "city"    => "Toronto",
            "state"   => "CA",
            "zip"     => "M5H2M5",
            "phone"   => "+1 416-360-6776",
            "country" => "CA"
        ),
        "customs_info" => $customs_info,
        "reference"    => "timhortons",
        "carrier"      => "USPS",
        "service"      => "ExpressMailInternational"
    )
);

// get a list of all my batches
// $all = \EasyPost\Batch::all();
// print_r($all);

// retrieve a batch
// $batch = \EasyPost\Batch::retrieve('batch_0SLoY64K');

// create shipment batch
$shipments = array();
for ($i = 0, $j = count($orders); $i < $j; $i++) {
    $shipments[] = array(
        "to_address"   => $orders[$i]["address"],
        "from_address" => $from_address,
        "parcel"       => $parcel,
        "customs_info" => $orders[$i]["customs_info"],
        "reference"    => $orders[$i]["reference"],
        "carrier"      => $orders[$i]["carrier"],
        "service"      => $orders[$i]["service"]
    );
}

$batch = \EasyPost\Batch::create(array('shipments' => $shipments));

// asynchronous creation means you can send us up to
// 1000 shipments at once, but you'll have to wait
// for the shipments to be created before you can continue
while ($batch->status->created != count($orders)) {
    sleep(5);
    $batch->refresh();
    if ($batch->status->creation_failed != 0) {
        throw new \EasyPost\Error('One of your batch shipments was unable to be created. Please manually retrieve and review your batch.');
    }
}

// if creation_failed == 0 and the while loop above ends
// we know that all shipments have been created
$batch->buy();

// asyncronous purchasing means we have to watch
// for when all labels have been purchased
while ($batch->status->postage_purchased != count($orders)) {
    sleep(5);
    $batch->refresh();
    if ($batch->status->postage_purchase_failed != 0) {
        throw new \EasyPost\Error('One of your batch shipments was unable to be purchased. Please manually retrieve and review your batch.');
    }
}

// generate a consolidated file containing all batch labels
$batch->label(array("file_format" => "pdf"));

while (empty($batch->label_url)) {
    sleep(5);
    $batch->refresh();
}

print_r($batch);


// request a scan form
$batch->create_scan_form();

// wait for scan form to complete
while (empty($batch->scan_form)) {
    sleep(5);
    $batch->refresh();
}

// retrieve scan form
$scan_form = \EasyPost\ScanForm::retrieve($batch->scan_form->id);

print_r($scan_form);
